### PR TITLE
[Refactor] improved ParticipationResource

### PR DESCRIPTION
**Pull request by PatZintel**
### Checklist
- [ ] I tested *all* changes and *all* related features with different users (student, tutor, editor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] Server: I implemented the changes with a good performance and prevented too many database calls
- [x] Server: I documented the Java code using JavaDoc style.

### Motivation and Context

adapt ParticipationResource to adhere to new guidelines of REST Resources

### Steps for Testing

1. start your participation in an exercise
2. update your participation by giving a participation point as tutor
3. look at a graded participation
4. delete a participation as instructor
5. delete the build plan for a programming exercise participation as instructor

### Review Progress

- Code Review
  - [ ] Review 1
  - [ ] Review 2
- Manual Tests
  - [ ] Test 1
  - [ ] Test 2

### Test Coverage

ParticipationResource: 93%

### CODE REVIEW

File: src/main/java/de/tum/in/www1/artemis/repository/TeamRepository.java [hunk 1]
```diff
@@ -99,6 +102,12 @@ default Team save(Exercise exercise, Team team) {
         return save(team);
     }
 
+    @NotNull
+    default Team findOneByExerciseIdAndUserIdElseThrow(Long exerciseId, Long userId) {
+        return findOneByExerciseIdAndUserId(exerciseId, userId)
+                .orElseThrow(() -> new EntityNotFoundException("Couldn't an active team for exercise " + exerciseId + " with user " + userId + "!"));
```
Comment (FrankeLukas) [inline src/main/java/de/tum/in/www1/artemis/repository/TeamRepository.java:108]:
```suggestion
                .orElseThrow(() -> new EntityNotFoundException("Couldn't find an active team for exercise " + exerciseId + " with user " + userId + "!"));
```

File: src/main/webapp/app/exercises/shared/participation/participation.service.ts [hunk 2]
```diff
@@ -87,9 +87,9 @@ export class ParticipationService {
     }
 
     cleanupBuildPlan(participation: StudentParticipation): Observable<EntityResponseType> {
-        const copy = this.convertDateFromClient(participation);
+        this.convertDateFromClient(participation);
```
Comment (FrankeLukas) [inline src/main/webapp/app/exercises/shared/participation/participation.service.ts:90]: If we don't send this via the network, we also don't need to convert anything, right?

File: src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseTestService.java [hunk 3]
```diff
@@ -647,8 +647,7 @@ public void startProgrammingExercise_correctInitializationState(ExerciseMode exe
             participant = setupTeam(user);
         }
         mockDelegate.mockConnectorRequestsForStartParticipation(exercise, participant.getParticipantIdentifier(), participant.getParticipants(), true, HttpStatus.CREATED);
-        final var path = ParticipationResource.Endpoints.ROOT + ParticipationResource.Endpoints.START_PARTICIPATION.replace("{courseId}", String.valueOf(course.getId()))
-                .replace("{exerciseId}", String.valueOf(exercise.getId()));
+        final var path = ParticipationResource.Endpoints.ROOT + '/' + ParticipationResource.Endpoints.START_PARTICIPATION.replace("{exerciseId}", String.valueOf(exercise.getId()));
```
Comment (FrankeLukas) [inline src/test/java/de/tum/in/www1/artemis/programmingexercise/ProgrammingExerciseTestService.java:650]: What is the difference between `ParticipationResource.Endpoints.ROOT` and just `ROOT`? Can we use `ROOT` for consistency?

**Review by FrankeLukas — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 4]
```diff
@@ -349,12 +345,12 @@ else if (exercise instanceof ModelingExercise modelingExercise) {
     }
 
     /**
-     * GET /participations/:participationId : get the participation for the given "participationId" including its latest result.
+     * GET participations/:participationId : get the participation for the given "participationId" including its latest result.
```
Comment (imstreetspirit) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:348]:
```suggestion
     * GET participations/:participationId/withLatestResult : get the participation for the given "participationId" including its latest result.
```

**Review by imstreetspirit — COMMENTED** [review]
just left very small comment, and agree with the suggestions mentioned by Lukas above.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 5]
```diff
@@ -616,9 +605,6 @@ private Course findCourseFromParticipation(StudentParticipation participation) {
 
         public static final String ROOT = "/api";
 
-        public static final String START_PARTICIPATION = "/courses/{courseId}/exercises/{exerciseId}/participations";
-
-        private Endpoints() {
-        }
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:622]: I would recommend keeping this (make it not instantiable).

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 6]
```diff
@@ -152,8 +150,7 @@ public ParticipationResource(ParticipationService participationService, Programm
 
         // if this is a team-based exercise, set the participant to the team that the user belongs to
         if (exercise.isTeamMode()) {
-            participant = teamRepository.findOneByExerciseIdAndUserId(exercise.getId(), user.getId())
-                    .orElseThrow(() -> new BadRequestAlertException("Team exercise cannot be started without assigned team.", "participation", "cannotStart"));
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:156]: Are we certain this alert is still displayed properly? I'm also not sure whether `Not Found` is better in this context than `Bad Request`. We require a team to be assigned, which does not exist then, but the problem does not seem to me like a "classic" missing entity, the exercise is found and a prerequisite not fulfilled. The opinion of other's would be useful here as well.

File: src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java [hunk 7]
```diff
@@ -213,14 +209,14 @@ private void addLatestResultToParticipation(StudentParticipation participation)
     }
 
     /**
-     * PUT /participations : Updates an existing participation.
+     * PUT exercises/{exerciseId}/participations : Updates an existing participation.
```
Comment (MaisiKoleni) [inline src/main/java/de/tum/in/www1/artemis/web/rest/ParticipationResource.java:212]: Are the braces intentional here? Or do we use `:`?

**Review by MaisiKoleni — CHANGES_REQUESTED** [review]
Changes look good to me in general, the renaming and de-duplication is great, I'm just not sure about the failed team participation start response.

This also needs extensive manual testing of all connected functionality, as a pure code review to check the matching client-server REST services is too hard. Maybe we find a better automatic way to check that as well.

**Comment by krusche** [comment]
This PR has too many merge conflicts!

**Pull request closed without merging**